src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.cc

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <libbson-1.0/bson.h>
+
+#include <rapidjson/document.h>
+#include <string>
+
+#include "src/common/base/base.h"
+#include "src/common/base/utils.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/types.h"
+#include "src/stirling/utils/binary_decoder.h"
+
+namespace px {
+namespace stirling {
+namespace protocols {
+namespace mongodb {
+
+ParseState ProcessOpMsg(BinaryDecoder* decoder, Frame* frame) {
+  PX_ASSIGN_OR(uint32_t flag_bits, decoder->ExtractLEInt<uint32_t>(), return ParseState::kInvalid);
+
+  // Find relavent flag bit information and ensure remaining bits are not set.
+  // Bits 0-15 are required and bits 16-31 are optional.
+  for (int i = 0; i <= 16; i++) {
+    if (i == 0 && (flag_bits >> i) & 1) {
+      frame->checksum_present = true;
+    } else if (i == 1 && (flag_bits >> i) & 1) {
+      frame->more_to_come = true;
+    } else if (i == 16 && (flag_bits >> i) & 1) {
+      frame->exhaust_allowed = true;
+    } else if (flag_bits >> i) {
+      return ParseState::kInvalid;
+    }
+  }
+
+  // Determine the number of checksum bytes in the buffer.
+  uint8_t checksum_bytes = (frame->checksum_present) ? 4 : 0;
+
+  // Get the section(s) data from the buffer.
+  while (decoder->BufSize() > checksum_bytes) {
+    mongodb::Section section;
+    PX_ASSIGN_OR(section.kind, decoder->ExtractLEInt<uint8_t>(), return ParseState::kInvalid);
+    // Length of the section still remaining in the buffer.
+    int32_t remaining_section_length = 0;
+
+    if (section.kind == 0) {
+      section.length = utils::LEndianBytesToInt<int32_t, 4>(decoder->Buf());
```

Comment:
The reason I didn't use the binary decoder here was because I didn't want to extract the length bytes out of the buffer since the `bson_new_from_data()` function expects the length bytes to be included in the data as well as the expected length of the data

Kind 1 sections have 3 fields (kind, section length, identifier) before the series of documents, so I was able to extract the section length bytes using the binary decoder but still had to use  `utils::LEndianBytesToInt<int32_t, 4>(decoder->Buf())`to find the document length at the time of BSON to JSON conversion

Reply:
Cool, that sounds like a valid reason for opting out of consistency. Can we add a comment here to explain that?